pheps/phep-0003.md

```diff
@@ -0,0 +1,356 @@
+```
+PHEP: 3
+Title: PyHC Python & Upstream Package Support Policy
+Author: Shawn Polson <pyhc-confidential@lasp.colorado.edu> <https://orcid.org/0000-0003-0619-5745>
+Discussions-To: https://github.com/heliophysicsPy/standards/pull/29
+Revision: 1
+Status: Draft
+Type: Standards Track
+Content-Type: text/markdown; charset=UTF-8; variant=CommonMark
+Created: 06-Jun-2024
+Post-History: 06-Jun-2024, 11-Jun-2024, 02-Jul-2024
+```
+
+# Abstract
+<a name="abstract"></a>
+This PHEP recommends that all projects across the PyHC ecosystem adopt a common time-based policy for support of dependencies, similar to [SPEC 0](https://scientific-python.org/specs/spec-0000/). Specifically, for Python versions and the upstream Scientific Python packages covered by SPEC 0, it recommends that projects:
+1. Support Python versions for at least **36 months** (3 years) after their initial release.
+2. Support upstream Scientific Python packages for at least **24 months** (2 years) after their initial release.
+3. Adopt support for new versions of these dependencies within **6 months** of their release.
+
+The upstream Scientific Python packages are: `numpy, scipy, matplotlib, pandas, scikit-image, networkx, scikit-learn, xarray, ipython, zarr`.
+
+This policy will replace the current standard [#11](https://github.com/heliophysicsPy/standards/blob/main/standards.md#standards) which simply mandates Python 3 support.
+
+# Motivation
+<a name="motivation"></a>
+The current PyHC standard [#11](https://github.com/heliophysicsPy/standards/blob/main/standards.md#standards), which mandates compatibility with Python 3, is outdated. Python 3 support is virtually universal now, so it would be more beneficial to replace this standard with a policy for how to support new minor Python versions and key upstream dependencies. [SPEC 0](https://scientific-python.org/specs/spec-0000/) provides a structured support timeline that balances stability and progress, essential for software in the heliophysics community. Adopting a similar policy ensures consistency and predictability in support timelines. Additionally, limiting the scope of supported versions is an effective way for packages to limit maintenance burden while promoting interoperability.
+
+# Rationale
+<a name="rationale"></a>
+Following [SPEC 0](https://scientific-python.org/specs/spec-0000/)'s 24/36-month support timeline keeps PyHC in better sync with the broader Scientific Python community, maintaining compatibility with newer Python features and key upstream dependencies, while providing adequate time for package maintainers to adapt. Allowing 6 months to adopt new versions ensures packages stay current with development cycles while providing a reasonable timeframe for testing and integration.
```

Comment:
I think it should be worth adding to the rationale that reducing the number of versions supported by projects reduces the maintenance burden. It's almost impossible to test oldest supported versions of all packages in a pre-SPEC 0 land.

SunPy basically used to bump our minimum versions when our CI failed, since adopting NEP 29 / SPEC 0 we haven't had to do that at all. So we have a lot more confidence that our code actually works with a concrete set of minimum deps which are clear to everyone.

Reply:
> I think it should be worth adding to the rationale that reducing the number of versions supported by projects reduces the maintenance burden.

💯 Absolutely right! This is personally my main motivation for adopting SPEC 0.

> It's almost impossible to test oldest supported versions of all packages in a pre-SPEC 0 land.

🦀 `uv` has [resolution strategies](https://github.com/astral-sh/uv?tab=readme-ov-file#resolution-strategy) for `--lowest` and `--lowest-direct` which makes it pretty straightforward nowadays to test against the oldest supported versions of either all dependencies or direct dependencies.  (We use `--lowest-direct` for PlasmaPy because some dependencies don't list minimum required versions of their dependencies.)

> SunPy basically used to bump our minimum versions when our CI failed, since adopting NEP 29 / SPEC 0 we haven't had to do that at all. So we have a lot more confidence that our code actually works with a concrete set of minimum deps which are clear to everyone.

Adopting NEP 29 / SPEC 0 has worked really well for PlasmaPy too!

Reply:
The last sentence of the `Motivation` section immediately preceding this `Rationale` one is: 

> `Additionally, limiting the scope of supported versions is an effective way for packages to limit maintenance burden while promoting interoperability.`

We probably don't need to say that twice in back-to-back sections? @Cadair @namurphy would you prefer I move that sentence from `Motivation` to `Rationale`? Or leave it as-is? Or would it actually be better to say that in both sections?

Reply:
Good point!  Personally, I'll be happy with it so long as it's mentioned in at least one place in the document.

Reply:
My recent push kept it in the same place. @Cadair if you'd like me to move it under `Rationale` instead I happily will.